cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/RunRecordCorrectorServiceTest.java

```diff
@@ -52,94 +54,141 @@
-    store.setRunning(wordcountFlow1.toEntityId(), rr.getPid(), nowSecs + 1, null, ByteBuffer.allocate(1).array());
-
-    // Now check again via Store to assume data store is wrong.
-    RunRecord runRecordMeta = store.getRun(wordcountFlow1.toEntityId(), rr.getPid());
-    Assert.assertNotNull(runRecordMeta);
-    Assert.assertEquals(ProgramRunStatus.RUNNING, runRecordMeta.getStatus());
-
-    // Verify there is NO FAILED run record for the application
-    runRecords = getProgramRuns(wordcountFlow1, ProgramRunStatus.FAILED);
-    Assert.assertEquals(0, runRecords.size());
-
-    // Start the RunRecordCorrectorService, which will fix the run record
-    CConfiguration testConf = CConfiguration.create();
-    // set threshold to 0 so that it will actually correct the record
-    testConf.set(Constants.AppFabric.PROGRAM_MAX_START_SECONDS, "0");
-    new LocalRunRecordCorrectorService(testConf, store, programStateWriter, programLifecycleService,
-                                       runtimeService).startUp();
-
-    // Wait for the FAILED run record for the application
-    Tasks.waitFor(1, new Callable<Integer>() {
-                    @Override
-                    public Integer call() throws Exception {
-                      return getProgramRuns(wordcountFlow1, ProgramRunStatus.FAILED).size();
-                    }
-                  }, 30, TimeUnit.SECONDS, 1, TimeUnit.SECONDS);
+    };
+
+    // Create a run record fixer that only has the flow running.
+    // Set the start buffer time to -1 so that it fixes right away.
+    // Also use a small tx batch size to validate the batching logic.
+    RunRecordCorrectorService fixer = new RunRecordCorrectorService(store, programStateWriter,
+                                                                    programRuntimeService, -1L, 5) { };
+    fixer.fixRunRecords();
+
+    // Validates all expected states
+    for (Map.Entry<ProgramRunId, ProgramRunStatus> entry : expectedStates.entrySet()) {
+      validateExpectedState(entry.getKey(), entry.getValue());
+    }
+
+    // Remove the workflow from the running set and mark it as completed
+    runningSet.remove(workflowId.getParent());
+    store.setStop(workflowId.getParent(), workflowId.getRun(), System.currentTimeMillis(),
+                  ProgramRunStatus.COMPLETED, Bytes.toBytes(sourceId.getAndIncrement()));
+
+    fixer.fixRunRecords();
+
+    // Both the workflow and the MR in workflow should be changed to failed state
+    expectedStates.put(workflowId, ProgramRunStatus.COMPLETED);
+    expectedStates.put(mrInWorkflowId, ProgramRunStatus.FAILED);
+
+    // Validates all expected states again
+    for (Map.Entry<ProgramRunId, ProgramRunStatus> entry : expectedStates.entrySet()) {
+      validateExpectedState(entry.getKey(), entry.getValue());
+    }
+  }
+
+  private void validateExpectedState(ProgramRunId programRunId, ProgramRunStatus expectedStatus) {
```

Comment:
this test case should be a lot faster now, right? 👍

Reply:
yes and test more conditions.